pkg/approver/aws/csr.go

```diff
@@ -0,0 +1,184 @@
+			Type:    certificates.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: csrr.successMessage,
+		})
+		_, err = ar.kubeClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+		if err != nil {
+			return fmt.Errorf("error updating approval for csr: %v", err)
+		}
+		break
+	}
+
+	return nil
+}
+
+type recognizerFunc func(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool
+type csrRecognizer struct {
+	recognizers    []recognizerFunc
+	successMessage string
+}
+
+func (ar *Approver) recognizers() []csrRecognizer {
+	return []csrRecognizer{{
+		// ensures: node present their client cert, exist(running) in cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isSelfNodeClientCert, ar.isValidNode, ar.isValidASG},
+		successMessage: "kube-aws-approver approved self node client cert",
+	}, {
+		// ensures: node username has instance-id, presented instance-id match instance-id by aws for node name in running state, doesn't belong to cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isNodeClientCert, ar.isValidNewNode, ar.isValidASG},
+		successMessage: "kube-aws-approver approved new node client cert",
+	}}
+}
+
+func isNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isSelfNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidNewNode(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	idu := getInstanceIDFromUsername(csr.Spec.Username)
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	idn, err := ar.aws.instanceID(nn)
+	if err != nil {
+		return false
+	}
+	if idn != idu {
+		return false
+	}
+	_, err = ar.kubeClient.CoreV1().Nodes().Get(nn, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidNode(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	_, err := ar.aws.instanceID(nn)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidASG(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	asg, err := ar.aws.autoScalingGroupID(nn)
```

Comment:
Is this the only call that uses the AWS API? Maybe a good place to make configurable for a test?

```
func (ar *Approver) newRecognizer(autoScalingGroupID func(nodeID string) (string, error)) recognizerFunc {
    return func(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
        // handle CSR
    }
}
```

Reply:
all isValidNewNode, isValidNode, and isValidASG depend on aws api...